### PR TITLE
[SPARK-36011][SQL] Disallow altering permanent views based on temporary views or UDFs

### PR DESCRIPTION
**Pull request by roryqi**
### What changes were proposed in this pull request?
PR #15764 disabled creating permanent views based on temporary views or UDFs.  But AlterViewCommand didn't block temporary objects.

### Why are the changes needed?
More robust view canonicalization.

### Does this PR introduce _any_ user-facing change?
Yes, now if you alter a permanent view based on temporary views or UDFs, the operation will fail.

### How was this patch tested?
Add new unit tests.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@jerqi can you create a new JIRA instead of reusing SPARK-18217? The fixed versions would be different at least.

**Comment by jerqi** [comment]
> @jerqi can you create a new JIRA instead of reusing [SPARK-18217](https://issues.apache.org/jira/browse/SPARK-18217)? The fixed versions would be different at least.

OK, It's done

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala [hunk 1]
```diff
@@ -910,4 +910,20 @@ abstract class SQLViewSuite extends QueryTest with SQLTestUtils {
       }
     }
   }
+
+  test("SPARK-18217: Disallow altering permanent views based on temporary views") {
```
Comment (linhongliu-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala:914]: let's move this to `PersistedViewTestSuite`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 2]
```diff
@@ -258,7 +258,9 @@ case class AlterViewAsCommand(
   def markAsAnalyzed(): LogicalPlan = copy(isAnalyzed = true)
 
   override def run(session: SparkSession): Seq[Row] = {
-    if (session.sessionState.catalog.isTempView(name)) {
+    val isTemporary = session.sessionState.catalog.isTempView(name)
+    verifyTemporaryObjectsNotExists(session.sessionState.catalog, isTemporary, name, query)
```
Comment (linhongliu-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala:262]: there is a `verifyAutoGeneratedAliasesNotExists` in the latest master code. could you put them together?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala (hunk 1 above)
Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala:914]: and let's also fix the jira number

Comment (jerqi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: OK

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala (hunk 2 above)
Comment (jerqi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala:263]: OK

**Review by linhongliu-db — COMMENTED** [review]
lgtm, cc @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewTestSuite.scala [hunk 3]
```diff
@@ -465,4 +465,35 @@ class PersistedViewTestSuite extends SQLViewTestSuite with SharedSparkSession {
...
     }
   }
+
+  test("SPARK-36011: Disallow altering permanent views based on temporary views or UDFs") {
+    import testImplicits._
+    val table = "t"
+    withTable(table) {
+      (1 to 10).toDF("id").write.saveAsTable(table)
+      val view = "v1"
+      val tView = "v2"
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewTestSuite.scala:475]: nit: I feel it's clearer to just hardcode "t", "v1", "v2"

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 4]
```diff
@@ -259,7 +259,10 @@ case class AlterViewAsCommand(
...
```
Comment (linhongliu-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala:263]: let's remove `verifyAutoGeneratedAliasesNotExists(analyzedPlan, isTemporary = false, name)` in `alterTemporaryView ` function

Comment (jerqi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala:263]: OK

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewTestSuite.scala (hunk 3 above)
Comment (jerqi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewTestSuite.scala]: OK

**Comment by cloud-fan** [comment]
thanks, merging to master/3.2!

**Comment by cloud-fan** [comment]
@jerqi can you help to create backport PRs for 3.1 and 3.0? thanks!

**Comment by jerqi** [comment]
> @jerqi can you help to create backport PRs for 3.1 and 3.0? thanks!

Ok, it's my pleasure.

**Pull request closed without merging**